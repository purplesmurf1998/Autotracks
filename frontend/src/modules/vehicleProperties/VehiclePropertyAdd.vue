<template>
  <div>
    <CForm @submit.prevent="submitNewProperty">
      <CAlert v-if="!!errorMessage" color="danger">{{ errorMessage }}</CAlert>
      <CInput
        label="Header Name"
        id="header-name-txt"
        placeholder="Enter the label for the vehicle property"
        :value.sync="label"
      />
      <CSelect
        label="Input Type"
        id="input-type-cmb"
        :options="['Text', 'Number', 'Currency', 'Date', 'Options', 'List']"
        :value.sync="inputType"
        placeholder="Please select the input type"
      />
      <CRow>
        <CCol>
          <CSelect
            label="Value required?"
            id="is-required-cmb"
            :options="['Yes', 'No']"
            :value.sync="isRequired"
          />
        </CCol>
        <CCol>
          <CSelect
            label="Visible in inventory?"
            id="is-visible-cmb"
            :options="['Yes', 'No']"
            :value.sync="visible"
          />
        </CCol>
      </CRow>
      <CInput
        label="Options"
        id="header-name-txt"
        placeholder="ex. Showroom;Demoline;Garage"
        :value.sync="options"
        v-if="inputType == 'Options'"
      />
      <CTextarea
        label="Description (optional)"
        placeholder="Enter a description of the vehicle property"
        rows="5"
        :value.sync="description"
      />
      <CRow class="justify-content-center mt-2">
        <CButton
          color="primary"
          type="submit"
          :disabled="disableButtons"
          id="create-staff"
        >
          Create
        </CButton>
        <CButton
          class="ml-1"
          color="danger"
          :disabled="disableButtons"
          @click="setAddingVehicleProperty(false)"
        >
          Cancel
        </CButton>
      </CRow>
    </CForm>
  </div>
</template>

<script>
const axios = require("axios");

export default {
  name: "VehiclePropertyAdd",
  props: ["setAddingVehicleProperty", "addNewVehicleProperty"],
  data() {
    return {
      label: "",
      inputType: "",
      isRequired: "Yes",
      visible: "Yes",
      options: "",
      description: "",
      errorMessage: null,
      disableButtons: false,
    };
  },
  methods: {
    submitNewProperty() {
      // simple validation step
      if (this.label == "" || this.inputType == "") {
        this.showErrorMessage("Header Name and Input Type are required fields");
        return false;
      }

      // parse the options
      let options = [];
      if (this.inputType == "Options" && !this.options.includes(";")) {
        this.showErrorMessage(
          "Options input type must have more than one option and be separated by a semicolon (;)"
        );
        return false;
      } else if (this.inputType == "Options") {
        options = this.options.split(";");
      }

      // create the new property object
      const newProperty = {
        label: this.label,
        inputType: this.inputType,
        isRequired: this.isRequired == "Yes",
        visible: this.visible == "Yes",
        description: this.description,
        options,
      };

      // post the the new property to the API
      this.createNewProperty(newProperty);
    },
    createNewProperty(newProperty) {
      axios({
        method: "POST",
        url: `${this.$store.state.api}/dealerships/${this.$route.params.dealershipId}/vehicles/properties`,
        headers: {
          Authorization: `Bearer ${this.$store.state.auth.token}`,
        },
        data: newProperty,
      })
        .then((response) => {
          if (response.data.success) {
            this.addNewVehicleProperty(response.data.payload);
            this.setAddingVehicleProperty(false);
          }
        })
        .catch((error) => {
          console.log(error);
          this.showErrorMessage("Unable to create the vehicle property.");
        });
    },
    showErrorMessage(message) {
      this.errorMessage = message;
      setTimeout(() => {
        this.errorMessage = null;
      }, 5000);
    },
  },
};
</script>

<style>
</style>