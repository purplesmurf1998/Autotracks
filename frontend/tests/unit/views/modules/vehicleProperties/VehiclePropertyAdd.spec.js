import {createLocalVue, shallowMount} from '@vue/test-utils'
import CoreuiVue from '@coreui/vue-pro'
import Vuex from 'vuex'

import VehiclePropertyAdd from './../../../../../src/modules/vehicleProperties/VehiclePropertyAdd.vue'

const localVue = createLocalVue()
localVue.use(CoreuiVue)
localVue.use(Vuex)

const store = new Vuex.Store({
    state: {
      sidebarShow: 'responsive',
      sidebarMinimize: false,
      asideShow: false,
      darkMode: false
    }
  })

  describe('VehiclePropertyAdd.vue', () => {
    const wrapper = shallowMount(VehiclePropertyAdd, { 
      store, 
      localVue
    })
    it('has a name', () => {
      expect(VehiclePropertyAdd.name).toBe('VehiclePropertyAdd')
    })
    it('should render correct content', () => {
      expect(wrapper.find('#header-name-txt').exists())
      expect(wrapper.find('#input-type-cmb').exists())
      expect(wrapper.find('#create-staff').exists())
      expect(wrapper.find('#is-visible-cmb').exists())
      expect(wrapper.find('#is-required-cmb').exists())
    })
    test('renders correctly', () => {
      expect(wrapper.element).toMatchSnapshot()
    })
  })