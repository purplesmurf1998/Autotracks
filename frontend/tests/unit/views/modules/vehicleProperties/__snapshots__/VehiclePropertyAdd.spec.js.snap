// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VehiclePropertyAdd.vue renders correctly 1`] = `
<div>
  <cform-stub>
    <!---->
     
    <cinput-stub
      id="header-name-txt"
      label="Header Name"
      lazy="400"
      placeholder="Enter the label for the vehicle property"
      type="text"
      value=""
    />
     
    <cselect-stub
      id="input-type-cmb"
      label="Input Type"
      options="Text,Number,Currency,Date,Options,List"
      placeholder="Please select the input type"
      value=""
    />
     
    <crow-stub
      gutters="true"
      tag="div"
    >
      <ccol-stub
        tag="div"
      >
        <cselect-stub
          id="is-required-cmb"
          label="Value required?"
          options="Yes,No"
          value="Yes"
        />
      </ccol-stub>
       
      <ccol-stub
        tag="div"
      >
        <cselect-stub
          id="is-visible-cmb"
          label="Visible in inventory?"
          options="Yes,No"
          value="Yes"
        />
      </ccol-stub>
    </crow-stub>
     
    <!---->
     
    <ctextarea-stub
      label="Description (optional)"
      lazy="400"
      placeholder="Enter a description of the vehicle property"
      rows="5"
      value=""
    />
     
    <crow-stub
      class="justify-content-center mt-2"
      gutters="true"
      tag="div"
    >
      <cbutton-stub
        activeclass="router-link-active"
        color="primary"
        event="click"
        exactactiveclass="router-link-exact-active"
        id="create-staff"
        routertag="a"
        target="_self"
        type="submit"
      >
        
        Create
      
      </cbutton-stub>
       
      <cbutton-stub
        activeclass="router-link-active"
        class="ml-1"
        color="danger"
        event="click"
        exactactiveclass="router-link-exact-active"
        routertag="a"
        target="_self"
        type="button"
      >
        
        Cancel
      
      </cbutton-stub>
    </crow-stub>
  </cform-stub>
</div>
`;
